dsgrid/config/representative_period_time_dimension_config.py

```diff
@@ -44,10 +48,93 @@ def check_dataset_time_consistency(self, load_data_df):
             self._format_handler.list_expected_dataset_timestamps(self.model.ranges), load_data_df
         )
 
-    def convert_dataframe(self, df, project_time_dim):
-        # TODO: Create a dataframe with timestamps covering project_time_dim.model.ranges
-        # for all model_years in the df, then join the df to that dataframe on the time columns.
-        # Account for time zone.
+    def build_time_dataframe(self):
+        time_cols = self.get_timestamp_load_data_columns()
+        schema = StructType(
+            [StructField(time_col, IntegerType(), False) for time_col in time_cols]
+        )
+
+        model_time = self.list_expected_dataset_timestamps()
+        df_time = _get_spark_session().createDataFrame(model_time, schema=schema)
+
+        return df_time
+
+    # def build_time_dataframe_with_time_zone(self):
+    #     return self.build_time_dataframe()
+
+    def convert_dataframe(self, df=None, project_time_dim=None, time_zone_mapping=None):
+        # in spark.dayofweek: 1=Sunday, 7=Saturday
+        # in pandas.dt.day_of_week: 0=Monday, 6=Sunday
+        # the mapping is: pandas.dt.day_of_week = [(i+7-2)%7 for i in spark.dayofweek]
+
+        if project_time_dim is None:
+            return df
+        if (
+            project_time_dim.list_expected_dataset_timestamps()
+            == self.list_expected_dataset_timestamps()
+        ):
+            return df
+
+        time_cols = self.get_timestamp_load_data_columns()
+        ptime_col = project_time_dim.get_timestamp_load_data_columns()
+        assert len(ptime_col) == 1, ptime_col
+        ptime_col = ptime_col[0]
+
+        # get unique timezones
+        if time_zone_mapping is not None:
+            key = [col for col in time_zone_mapping.columns if col != "time_zone"]
+            df = df.join(time_zone_mapping, on=key)
+        geo_tz_values = [row.time_zone for row in df.select("time_zone").distinct().collect()]
+        geo_tz_names = [get_timezone(tz).tz_name for tz in geo_tz_values]
+
+        # create time map
+        # temporarily set session time to UTC for timeinfo extraction
+        # Note: timeinfo is extracted from local_time column exactly in DF.show(),
+        # and only UTC seems to convert to local_time correctly for DF.show().
+        # Even though UTC does not always lead to correct time output when DF.toPandas()
+        # the underlying time data is correctly stored when saved to file
```

Comment:
Is this still relevant?

Reply:
I think so, this is intended to explain why we are setting session tz to UTC temporarily for the time info extraction